lib/ecto/schema.ex

```diff
@@ -1329,17 +1329,22 @@ defmodule Ecto.Schema do
     end
   end
 
-  defp do_load(struct, types, map, loader) when is_map(map) do
+  def __load__(schema, prefix, source, context, data, loader) when is_map(data) do
+    __load__(schema, prefix, source, context, Map.to_list(data), loader)
```

Comment:
Sorry, I gave bad advice. We are not traversing the map, so keeping it as a map will be more efficient.